A few FV interfaces kernels

GiudGiud opened the pull request:
Refs #17638 

Early PR to get feedback on FVScalarLagranceMultiplierInterface

TODO

- [x] Find what s wrong with the FVScalarLagranceMultiplierInterface
- [x] Clean up tests
- [x] Figure out division by point then fix up convection and diffusion interface kernels
- [x] Look at dynamic number of ghosted cells

GiudGiud committed:
Add interface diffusive flux postprocessors, refs #17637

GiudGiud committed:
Add convection, diffusion, and continuity interface kernels, refs #17638
Add tests for each new interface kernels
Fixup interfaceDiffusiveFluxPostprocessor
Add test for interfaceDiffusiveFluxIntegral/Average

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/725390/) on 8d5770c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17640/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17640/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0c08243d3ae0bd6d64527bb2351f4a6c937d7c49`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/725417/) on 4ad31d5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17640/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17640/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0c08243d3ae0bd6d64527bb2351f4a6c937d7c49`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/725443/) on 98f2d6a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17640/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/725783/) on 05a605e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17640/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17640/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0c08243d3ae0bd6d64527bb2351f4a6c937d7c49`

lindsayad commented on framework/src/fviks/FVScalarLagrangeMultiplierInterface.C:
```diff
@@ -0,0 +1,123 @@
+
+  // LM residual. We may not have any actual ScalarKernels in our simulation so we need to manually
+  // make sure the scalar residuals get cached for later addition
+  const auto lm_r = MetaPhysicL::raw_value(computeQpResidual()) * fi.faceArea() * fi.faceCoord();
```
I'm guessing you have a more complex problem than this... one problem I see is that right now this would raise a `mooseError` :laughing:

GiudGiud replied:
lol I ll delete that definition. It's virtual though it's overwritten in the interface kernel I care about

GiudGiud replied:
ah it wasnt even virtual. Well now it's pure virtual so we are safe.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/726448/) on f109527 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17640/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17640/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0c08243d3ae0bd6d64527bb2351f4a6c937d7c49`

GiudGiud commented:
Would like a look on the need for ghosting for the bulk cell in the convection interface when you have time @lindsayad 
Can this be dynamic ?

lindsayad commented on modules/navier_stokes/doc/content/source/fviks/FVConvectionCorrelationInterface.h:
```diff
@@ -0,0 +1,13 @@
+# FVConvectionCorrelationInterface
```
Think you want this to be an `.md` file :-)

lindsayad commented on modules/navier_stokes/src/fviks/FVConvectionCorrelationInterface.C:
```diff
@@ -0,0 +1,87 @@
+ADReal
+FVConvectionCorrelationInterface::computeQpResidual()
+{
+  const Elem * current_elem = elemIsOne() ? &_face_info->elem() : _face_info->neighborPtr();
```
This variable name is not very clear. Both `elem()` and `neighborPtr()` are potentially "current" elements

GiudGiud replied:
right. I ll switch to elem_one

lindsayad reviewed (changes requested):
I enjoy the "wacky" new functions I get to see whenever I review a PR by you or @smharper. You guys and your real simulations :laughing: 

So off the top of my head, for doing this "cleanly" (in quotes because it's still not super clean feeling)... I would create a probably subdomain restricted user object that computes/holds "bulk" elements. I'm guessing that you want elements that run down some center-line for flow in pipes, maybe something more exotic for more complex geometries, but that's in the details. You should be able to query this user object for a bulk element best associated with data from your interface kernel. This user object would also do the following: either create a custom `RelationshipManager` (see `FVKernel::validParams()`) that ghosts center-line elements (or potentially more general than that), *or* uses TIMPI `push_pull_parallel_vector_data` that can snag data from other processors. @fdkong and @loganharbour are better references for that kind of thing

GiudGiud commented:
A bulkUO does sound tempting. It ll allow to define bulk not just based on distance from the wall, but like gradients of variables and such. 

I dont think I have the bandwidth for that right now though. And the cleaner way to handle convection will be with a wall function. So I ll just remove that option for now. The boolean should stay to force users to think about their y+ though

fdkong reviewed (changes requested):
It looks nice

lindsayad commented on framework/src/fviks/FVTwoVarContinuityConstraint.C:
```diff
@@ -0,0 +1,36 @@
+ADReal
+FVTwoVarContinuityConstraint::computeQpResidual()
+{
+  return var1().getBoundaryFaceValue(*_face_info) - var2().getBoundaryFaceValue(*_face_info);
```
This `computeQpResidual` is actually the residual for a Lagrange Multiplier so he is solving the constraint exactly (no penalty required)

GiudGiud commented on framework/src/fviks/FVDiffusionInterface.C:
```diff
@@ -0,0 +1,63 @@
+
+ADReal
+FVDiffusionInterface::computeQpResidual()
+{
```
yep

GiudGiud commented on framework/src/postprocessors/InterfaceDiffusiveFluxIntegral.C:
```diff
@@ -0,0 +1,114 @@
+                                        "primary side material property name is used for the "
+                                        "secondary side. Only needed for finite volume");
+  params.addClassDescription("Computes the diffusive flux on the interface.");
+  params.set<unsigned short>("ghost_layers") = 2;
```
for the gradient definition I think you're right

GiudGiud commented on framework/src/postprocessors/InterfaceDiffusiveFluxIntegral.C:
```diff
@@ -0,0 +1,114 @@
+    return -diffusivity * MetaPhysicL::raw_value(gradient * normal);
+  }
+  else
+    return -MetaPhysicL::raw_value(_diffusion_coef[_qp]) * _grad_u[_qp] * _normals[_qp];
```
yes

GiudGiud commented:
An alternative to using the direction between two point to support the gradient is to use the surface normal and use the gradient projection. 

What do you think is best here?

lindsayad commented on framework/src/fviks/FVDiffusionInterface.C:
```diff
@@ -0,0 +1,63 @@
+  const auto & coef_neighbor = elemIsOne() ? _coeff2_neighbor : _coeff1_neighbor;
+
+  // Form a finite difference gradient across the interface
+  Point du = _face_info->elem().centroid() - _face_info->neighborPtr()->centroid();
```
`centroid` calculations are somewhat expensive so we cache them upon `FaceInfo` creation in `elemCentroid()` and `neighborCentroid()` so let's make sure we leverage those.

lindsayad commented on framework/src/fviks/FVDiffusionInterface.C:
```diff
@@ -0,0 +1,63 @@
+
+  // Form a finite difference gradient across the interface
+  Point du = _face_info->elem().centroid() - _face_info->neighborPtr()->centroid();
+  du /= (du * du);
```
I find `du` confusing. I've grokked it now but it took me a few minutes. Could you name this variable differently? E.g. something like `dxmag` and then do division by that value instead of multiplication?

GiudGiud replied:
I dont think I can divide by Point

lindsayad commented:
> An alternative to using the direction between two point to support the gradient is to use the surface normal and use the gradient projection.
> 
> What do you think is best here?

I personally like the two point. The more I've been reading, it seems like at least OpenFOAM *always* does two-point for diffusion operators which helps limit the stencil and consequently any Jacobian storage in the case we do an implicit solve.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/738232/) on 8eaca7c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17640/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17640/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1fcc8c77d72b1faf1997c604efd5fe17f1cab86c`

GiudGiud committed:
Add more documentation, clang fix, refs #17638

GiudGiud commented:
Ok this should be ready for review / merge (let s see how it fares on the tests)
Since the number of ghost layers can be set in the inputs, i m actually happy enough with leaving the bulk_distance parameter. 
One question though, can i improve on the check for 'this point is in the local domain' in the replicated mesh case where the failure will not be caught by checking the output of the point locator ?

GiudGiud commented:
Also just checking, is there always at least one ghost layer on FVIks ?

loganharbour commented:
> Also just checking, is there always at least one ghost layer on FVIks ?

You get the ghosting with FVKernels (see [here](https://github.com/idaholab/moose/blob/next/framework/src/fvkernels/FVKernel.C#L34)), but we should probably add RMs to FVIKs as well.

GiudGiud commented:
![image](https://user-images.githubusercontent.com/25037151/117698990-8d08e500-b181-11eb-82bb-1bf75a70e79e.png)
Just a picture to feature the convection interface kernel (high res of the channel test)

I ll address the test failures soon. There's a syntax one, I think I should address the source of that first

GiudGiud committed:
Address test failures:
-move assert on residual size
-dont use class description for a class that is not registered / not meant for input files
-point to the right input file in documentation
-dont use local ad indexing for the interface kernel tests

Avoid seg fault when debugging BCs
Add ghost layers to FV interface kernels

lindsayad commented on framework/include/postprocessors/InterfacePostprocessor.h:
```diff
@@ -15,7 +15,7 @@
 /**
  * BASE CLASS FOR IMPLEMENTING INTERFACE POSTRPOCESSOR
  * ALL INTERFACE POSTPROCESOR SHOULD DERIVE FROM THIS CLASS
- * THIS CALSS ALASO COMPUTE THE INTERFACE AREA
+ * THIS ALSO COMPUTES THE INTERFACE AREA
```
WHY ARE YOU SHOUTING AT ME! Haha

lindsayad commented on framework/src/fviks/FVInterfaceKernel.C:
```diff
@@ -49,6 +49,24 @@ FVInterfaceKernel::validParams()
+  // FV Interface Kernels always need one layer of ghosting because the elements
+  // on each side of the interface may be on different MPI ranks, but we still
+  // need to access them as a pair to compute the numerical face flux.
+  params.addRelationshipManager(
+      "ElementSideNeighborLayers",
+      Moose::RelationshipManagerType::GEOMETRIC | Moose::RelationshipManagerType::ALGEBRAIC |
+          Moose::RelationshipManagerType::COUPLING,
+      [](const InputParameters & obj_params, InputParameters & rm_params) {
+        rm_params.set<unsigned short>("layers") = obj_params.get<unsigned short>("ghost_layers");
+        rm_params.set<bool>("use_point_neighbors") = obj_params.get<bool>("use_point_neighbors");
+      });
```
:+1:

lindsayad reviewed (changes requested):
This looks great to me other than the `friend class` stuff. Maybe we should just make some of the data protected. If you found it useful to be able to use the data here then other `FVInterfaceKernel` developers also might find it useful.

GiudGiud commented on framework/include/fviks/FVInterfaceKernel.h:
```diff
@@ -156,4 +161,6 @@ class FVInterfaceKernel : public MooseObject,
 
   const MooseMesh & _mesh;
+
+  friend class FVScalarLagrangeMultiplierInterface;
```
Yeah I was getting around the private-ness.

Ok I ll update thanks!

GiudGiud committed:
Address Alex's review, dont use friend class but use accessors instead, refs #17638

lindsayad commented on framework/include/fviks/FVInterfaceKernel.h:
```diff
@@ -123,6 +123,11 @@ class FVInterfaceKernel : public MooseObject,
+  /**
+   * @return The underlying assembly object, to work on residuals for example
+   */
+  Assembly & assembly() const { return _assembly; }
```
There’s not much point to this. You’re returning a non const reference so this does not provide any increased encapsulation relative to having the data member protected.

lindsayad replied:
It’s also misleading to label the method as const when you’re returning non const data

GiudGiud replied:
ok changed it.
I was doing this to avoid the re-ordering issue. I think the current re-ordering is ok?

GiudGiud committed:
Re-organize declarations for FVinterfaceKernels, refs #17638

GiudGiud commented:
@fdkong have the changes been addressed? Looking to merge this

fdkong commented:
Let me take a quick check. I am doing that now

fdkong commented:
Some failings in next 

https://civet.inl.gov/job/742044/

GiudGiud commented:
Seems to be some high-ish but not too high distributed cases for the interface diffusive flux postprocessor. 

It works fine on my machine though. Even in debug mode
